### ISSUE TITLE
Set() call ignores ES6 class instance properties

### ISSUE THREAD

**Issue by ulrichb** [Automattic/mongoose#7639]
**Do you want to request a *feature* or report a *bug*?** Bug

**What is the current behavior?**

The following TypeScript example (but the same applies to pure ES6 classes) shows that mongoose ignores `Nested.prop` in the `set()` call. => So the `nested`-property update gets lost in a following `save()` call.

```TS
class Nested {
    constructor(readonly prop: string) { }
}

const ClassInstanceAsNestedInputModel = mongoose.model<{ nested: Nested } & Document>(
    'ClassInstanceAsNestedInput', new Schema({
            nested: { prop: { required: true, type: String } },
        },
        // { strict: "throw" },
    ));

let doc = new ClassInstanceAsNestedInputModel({ nested: { prop: "initial" } });

const newNested = new Nested("new val");
console.log(newNested); // prints "Nested { prop: 'new val' }"

doc.set({ nested: newNested });
//doc.set({ nested: { ...newNested } }); // Workaround via object spread

// ISSUE: prints "{ prop: 'initial' }" instead of "{ prop: 'new val' }"
console.log(doc.nested);

// With 'strict: "throw"', set() throws "ObjectExpectedError: Tried to set nested
// object field `nested` to primitive value `[object Object]`"
```

I would classify this as bug because if someone does not use `{ strict: "throw" }` (which isn't the default) these "lost updates" aren't detected (i.e. a following `save()` stores a stale version of `nested`).

**What is the expected behavior?**
`doc.nested.prop` is "new val" after the `set()` call, i.e. `save()` (and all the other operations which "parse" a document input) also respect class properties.

**What are the versions of Node.js, Mongoose and MongoDB you are using?** 
Mongoose 5.4.19, 5.4.20, Node 11.12

**Comment by jglcanalcar** [comment]
We have detected this error too making a document.set.

The error has been detected in the commit bb78de76e10faf7514fefe423dfcdf5d2e2ec215

**Comment by ulrichb** [comment]
Hi @jglcanalcar,

What do you mean with "the error has been _detected_ in the commit bb78de7"?

I've just tested that the issue above still exists in 5.4.20 (which includes bb78de7) and also existed in 5.4.19.

**Comment by jglcanalcar** [comment]
@ulrichb I mean that we found that in production we had 5.4.20 and in a very business operation we were having several problems, meanwhile in 5.4.14 we didn't, so I cloned the repo, removed my node_modules/mongoose directory, and linked in my node_modules directory the cloned repo.

Then using git bisect I set as the bad commit the last one in the master branch, and the good one the one in the tag 5.4.11. Git bisect moves between commits doing binary search, so I tested several versions (commits) until I saw that the commit that introduced the bug was bb78de7, which I don't know in which TAG version it belongs to.
